OpenVPN master additions

julek-wolfssl opened the pull request:
- Implement RFC 5705: Keying Material Exporters for TLS
- SHA1 should not be called SHA. Leave alias for compatibility
- Fix call to wolfSSL_connect when in wolfSSL_connect_TLSv13
- Generate correct GCM tag
- Check that curves in set_groups functions are valid

julek-wolfssl committed:
Implement missing functionality for OpenVPN 2.5

julek-wolfssl committed:
Set options when creating SSL

julek-wolfssl committed:
Fix getting cipher suites in compat layer

julek-wolfssl committed:
Add static buffer to wolfSSL_ERR_error_string

Add ED448 and ED25519 to wolfssl_object_info
Add more error messages

julek-wolfssl committed:
Implement RFC 5705: Keying Material Exporters for TLS

julek-wolfssl committed:
Check that curves in set_groups functions are valid

julek-wolfssl committed:
SHA1 should not be called SHA. Leave alias for compatibility.

julek-wolfssl committed:
GCC complains about empty if

julek-wolfssl committed:
Fix call to wolfSSL_connect when in wolfSSL_connect_TLSv13

If a client is:
- TLS 1.3 capable
- calls connect with wolfSSL_connect_TLSv13
- on an WOLFSSL object that allows downgrading
then the call to wolfSSL_connect should happen before changing state to HELLO_AGAIN. Otherwise wolfSSL_connect will assume that messages up to ServerHelloDone have been read (when in reality only ServerHello had been read).

Enable keying material for OpenVPN

julek-wolfssl committed:
Rebase fixes

- wolfSSL_CTX_set1_groups_list and wolfSSL_set1_groups_list should use wolfSSL_CTX_set1_groups and wolfSSL_set1_groups respectively because it converts to correct groups representation
- Change to using "SHA1" as main name for SHA1

julek-wolfssl committed:
Detect version even if not compiled in

julek-wolfssl committed:
Generate correct GCM tag

SparkiDev commented on src/ssl.c:
```diff
@@ -46581,6 +46745,36 @@ static WC_INLINE int SCSV_Check(byte suite0, byte suite)
+        return 1;
+    if (cipher_names[i].minor < ssl->options.minDowngrade)
+        return 1;
+    switch (cipher_names[i].minor) {
```
TLS 1.0 cipher suites are available in TLS 1.1 and 1.2 but not 1.3.
Does this code need to handle these cases?

julek-wolfssl replied:
This is only used in `wolfSSL_get_ciphers_compat` so it is used to return the available ciphersuites to the user. I changed this to:
```
    if (cipher_names[i].minor < ssl->options.minDowngrade) {
        if (ssl->options.minDowngrade <= TLSv1_2_MINOR &&
                cipher_names[i].minor >= TLSv1_MINOR)
            /* 1.0 ciphersuites are in general available in 1.1 and
             * 1.1 ciphersuites are in general available in 1.2 */
            return 0;
        return 1;
    }
```
This will show ciphers as available if the min version is <= 1.2

SparkiDev commented on tests/api.c:
```diff
@@ -4528,7 +4528,8 @@ static void test_wolfSSL_UseTruncatedHMAC(void)
 static void test_wolfSSL_UseSupportedCurve(void)
 {
-#if defined(HAVE_SUPPORTED_CURVES) && !defined(NO_WOLFSSL_CLIENT) && !defined(NO_TLS)
+#if defined(HAVE_ECC) && defined(HAVE_SUPPORTED_CURVES) && \
```
ECC only when TLS 1.2 or lower.

SparkiDev commented on wolfssl/ssl.h:
```diff
@@ -1057,6 +1063,20 @@ WOLFSSL_API int  wolfSSL_CTX_get_cert_cache_memsize(WOLFSSL_CTX*);
 WOLFSSL_API int  wolfSSL_set_cipher_list(WOLFSSL*, const char*);
 
+#ifdef HAVE_KEYING_MATERIAL
+#define TLS_PRF_LABEL_CLIENT_FINISHED     "client finished"
```
Why are these public?

julek-wolfssl replied:
They don't need to be. I'll move them to ssl.c.

julek-wolfssl committed:
Jenkins Fixes

- explicit conversions
- not all curves available for wolfSSL_CTX_set1_groups_list
- group funcs depend on HAVE_ECC
- `InitSuites` after `ssl->suites` has been set

julek-wolfssl commented on src/ssl.c:
```diff
@@ -16308,33 +16464,21 @@ int wolfSSL_CTX_set_min_proto_version(WOLFSSL_CTX* ctx, int version)
     }
 
     switch (version) {
-#if defined(WOLFSSL_ALLOW_SSLV3) && !defined(NO_OLD_TLS)
```
I'll change it back to how it was and have OpenVPN testing require the additional TLS compilation options.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -2405,11 +2403,17 @@ int wolfSSL_UseSupportedCurve(WOLFSSL* ssl, word16 name)
         case WOLFSSL_FFDHE_6144:
         case WOLFSSL_FFDHE_8192:
-            break;
+            return 0;
```
`return 0` meant that it was a valid group. I changed it so that 1 is valid and 0 is invalid.

SparkiDev replied:
You extracted this code toa new function that returns a boolean. Yes?

julek-wolfssl replied:
@toddouska @SparkiDev The return code was not changed. I refactored the curve validity checking into a static function (for use in other group functions) that returns 1 on a valid curve and 0 on an invalid one. `wolfSSL_UseSupportedCurve` still has the exact same return code.
